src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -517,41 +522,31 @@ internal virtual void AddHeaders(HttpHeaders sourceHeaders)
             Debug.Assert(sourceHeaders != null);
             Debug.Assert(GetType() == sourceHeaders.GetType(), "Can only copy headers from an instance of the same type.");
 
-            Dictionary<HeaderDescriptor, object>? sourceHeadersStore = sourceHeaders._headerStore;
-            if (sourceHeadersStore is null || sourceHeadersStore.Count == 0)
-            {
-                return;
-            }
-
-            _headerStore ??= new Dictionary<HeaderDescriptor, object>();
-
-            foreach (KeyValuePair<HeaderDescriptor, object> header in sourceHeadersStore)
+            foreach (HeaderEntry entry in sourceHeaders.GetEntriesGroupedByName())
```

Comment:
I don't think we need to force grouping here. This function basically clones the headers from another collection (sourceHeaders). It seems like it's fine to clone them ungrouped. 

Note we also don't force headers to be parsed in this case.

Reply:
Imagine the source collection had entries like
```
Foo: a
Foo: b
```

With a loop like
```c#
foreach (HeaderEntry entry in source.GetEntries())
{
    if (!ContainsEntry(entry.Key))
    {
        Add(entry);
    }
}
```

You would end up adding `Foo: a`, then seeing that the key is already present and never add `Foo: b`.
If the collection has been transformed into a dictionary, we can't efficiently tell if the entry existed before we performed any `Add`s.

We could do something like https://github.com/MihaZupan/runtime-1/commit/32d01f98303a19cff831e453894a2ff84d155eee, but do you think it's worth it given that we already avoid grouping on a best-effort basis (we don't guarantee it)?
We also primarily avoid grouping for the performance benefit, but a previously grouped collection may become marked as ungrouped as part of adding the headers, potentially slowing down future operations.

Reply:
Thanks for the example -- I understand the issue here now. Your original approach seems fine. Would be useful to add a comment explaining this.